packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,205 @@
+import { SequenceChangeset } from "./sequenceChangeset";
+
+/**
+ * Composes a sequence of changesets into a single changeset.
+ * @param changes - The changesets to be applied.
+ * Each changeset in the list is assumed to be applicable after the previous one.
+ * @returns A changeset that is equivalent to applying each of the given `changes` in order.
+ *
+ * WARNING! This implementation is incomplete:
+ * - Tombstone information is ignored.
+ * - Support for moves is not implemented.
+ * - Support for slices is not implemented.
+ */
+export function compose(changes: SequenceChangeset[]): SequenceChangeset {
+    const base: SequenceChangeset = {
+        marks: {},
+    };
+    for (const change of changes) {
+        foldInFieldMarks(change.marks, base.marks);
+    }
+    return base;
+}
+
+function foldInFieldMarks(newFieldMarks: T.FieldMarks, baseFieldMarks: T.FieldMarks) {
+    for (const key of Object.keys(newFieldMarks)) {
+        const newMarkList = newFieldMarks[key];
+        baseFieldMarks[key] ??= [];
+        const baseMarkList = foldInMarkList(newMarkList, baseFieldMarks[key]);
+        baseFieldMarks[key] = baseMarkList;
+        while (typeof baseMarkList[baseMarkList.length - 1] === "number") {
+            baseMarkList.pop();
+        }
+        if (baseMarkList.length === 0) {
+            // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+            delete baseFieldMarks[key];
+        }
+    }
+}
+
+function foldInMarkList(
+    newMarkList: T.MarkList,
+    baseMarkList: T.MarkList,
+): T.MarkList {
+    const factory = new MarkListFactory();
+    let iBase = 0;
+    let iNew = 0;
+    let nextBaseMark: T.Mark | undefined = baseMarkList[iBase];
+    let nextNewMark: T.Mark | undefined = newMarkList[iNew];
+    while (nextNewMark !== undefined) {
+        let newMark: T.Mark = nextNewMark;
+        let baseMark: T.Mark = nextBaseMark;
+        nextNewMark = undefined;
+        nextBaseMark = undefined;
+        if (baseMark === undefined) {
+            factory.push(clone(newMark));
+        } else if (isAttachGroup(newMark) || isReattach(newMark)) {
+            factory.pushContent(clone(newMark));
+            nextBaseMark = baseMark;
+        } else if (isDetachMark(baseMark)) {
+            // TODO: match base detaches to tombs and reattach in the newMarkList
+            factory.pushContent(baseMark);
+            nextNewMark = newMark;
+        } else {
+            const newMarkLength = getMarkLength(newMark);
+            const baseMarkLength = getMarkLength(baseMark);
+            if (newMarkLength < baseMarkLength) {
+                [baseMark, nextBaseMark] = splitMark(baseMark, newMarkLength);
+            } else if (newMarkLength > baseMarkLength) {
+                [newMark, nextNewMark] = splitMark(newMark, baseMarkLength);
+            }
+            // Passed this point, we are guaranteed that `newMark` and `baseMark` have the same length
+            if (typeof baseMark === "number") {
```

Comment:
Seems weird that this case is handled separately, while `compose` handles the case where `newMark` is a `Skip`. Is this because of the remaining work described in the TODO?